src/passes/OptimizeInstructions.cpp

```diff
@@ -2557,6 +2557,21 @@ struct OptimizeInstructions
         return inner;
       }
     }
+    {
+      // (i32(x) < 0) | (i32(y) < 0)   ==>   i32(x | y) < 0
+      // (i64(x) < 0) | (i64(y) < 0)   ==>   i64(x | y) < 0
```

Comment:
Another approach here might be to turn `x <(signed) 0  =>  x >>> 31`, as it's basically a check for the sign bit. I'm not sure if that's better, though.

If it makes sense, then
```
(x < 0) | (y < 0)  =>  (x >>> 31) | (y >>> 31)  => (x | y) >>> 31
```
The last change could use a more generalized rule,
```
(x >>> K) | (y >>> K)  => (x | y) >>> K
```

Reply:
I would like to avoid `x < 0` conversion here. I proposed canonicalized `x >>> 31` into `x < 0` and back as finlize optimization here: https://github.com/WebAssembly/binaryen/pull/4160

Reply:
But `x >> z | y >> z` could be generalized and handled by one-or-two generic rule. I created snipped for the here: https://github.com/WebAssembly/binaryen/issues/1764#issuecomment-972746883

Reply:
Here new PR with low priority: https://github.com/WebAssembly/binaryen/pull/4342.

Reply:
Right, probably `x >>> 31` doesn't make sense to do in general. 31 is less common than 0 so it would compress less well, I'd bet.

I do still feel there is maybe something more generic here. `F(x) | F(y)  =>  F(x | y)` is true for various `F` and not just `F(x) := x < 0`. Perhaps we can find the right conditions for `F`?

Reply:
`F` should be definitely binary expression. And should have a lot of complicated constraints. For example comparison operators may reduced in only limit cases like this. This example `x < 2 | y < 2` already not satisfy constrain and rest of constant as well. Another thing #4342, but even there are exceptions. Perhaps some SAT solver could find some formalized subset, but I'm not sure it would be global.

Reply:
I wouldn't suggest using a SAT solver, yeah, that would be far too complicated. But I'm suggesting we spend some time to try to find some simple property here.

At minimum, "F just shifts bits around" is a fairly general property, which would also handle the current PR, but hopefully we can do better.

Reply:
We can of course try to find a more generalized rule especially for unary operations. Perhaps even for some binary operations like shifts. But this current of relational operations with constant on LHS will be a special edge case in any way. Or am I missing something?

Reply:
We can't do this for example:
```rust
(x + z) | (y + z)   ->   (x | y) + z   or  any other reductions
(x - z) | (y - z)   ->   (x | y) - z   or  any other reductions
(x < z) | (y < z)   ->   (x | y) < z   or  any other reductions
(x ^ z) | (y ^ z)   ->   (x | y) ^ z   or  any other reductions
```
all this wrong.

We could simplify only these:
```rust
(x & z) | (y & z)  ->  (x | y) & z
(x | z) | (y | z)  ->  (x | y) | z

But that rules could be part of different generic rewrite approach which reassociate such expressions first:
```rust       
(x | z) | (y | z)   ->   (x | y) | (z | z)     ->   (x | y) | z
```
For `(x & z) | (y & z)  ->  (x | y) & z` we could be create rule

Reply:
I wondered how the same GCC handles this. And in its rules these private rules are also described separately:
https://github.com/gcc-mirror/gcc/blob/master/gcc/match.pd#L937
https://github.com/gcc-mirror/gcc/blob/master/gcc/match.pd#L946

for at least `x == 0 | y == 0` and `x == -1 & y == -1`. The rest rules as well:
https://github.com/gcc-mirror/gcc/blob/master/gcc/match.pd#L1425
https://github.com/gcc-mirror/gcc/blob/master/gcc/match.pd#L1446
https://github.com/gcc-mirror/gcc/blob/master/gcc/match.pd#L1457

Reply:
But I found some complex general rule for `(x op1 C1) | (y op2 C2)` where `op1, op2` <-- `>`, `>=`, `<`, `<=`
https://github.com/gcc-mirror/gcc/blob/master/gcc/match.pd#L2406

Reply:
Maybe this is a larger topic, then. To not block this PR, how about this: We can define a function `isHomomorphicOnOr()` as `F(x) | F(y) == F(x | y)` is basically the definition of a mathematical homomorphism on the Or operation,

https://en.wikipedia.org/wiki/Homomorphism#Definition

However maybe the math term is too fancy, and we could just say `preservesOr()`. And then that function can for now just recognize the pattern you want here. But writing it with such a helper function would help us to do later refactorings and improvements.

Reply:
Well, all this very non-trivial and will required some kind of proof solver and some algebra like F-algebra or somthing like. For example LLVM can't do this. Instead it split `|` into large number of special cases and rules here:
https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp#L2541

Reply:
Sorry if I wasn't clear: All I am suggesting in this PR is to add a function `preservesOr()`, and to call that function. The function's body would just contain the one pattern so far that we know preserves the Or operation through it. Anything non-trivial would be left for the future.

Reply:
Perhaps I misunderstood. Okay. I put those rules in the preserveOr